File: burn.c
/* Small CPU-burning test program for exercising hamlet's userspace symbol
 * resolution. Runs a chain of nested function calls in a tight infinite
 * loop, so sampled stack traces show distinct frames: main -> foo -> bar.
 *
 * noinline + no optimization keep each function as a real, separate
 * symbol in the binary instead of being inlined away.
 *
 * Build (keep the symbol table - do NOT strip):
 *     gcc -O0 -fno-inline -o burn burn.c
 *
 * Run:
 *     ./burn &
 *     cargo run --release -- --pid $!
 */

__attribute__((noinline))
long bar(long x) {
    long total = 0;
    for (long i = 0; i < 1000; i++) {
        total += (x * i) % 7919;
    }
    return total;
}

__attribute__((noinline))
long foo(long x) {
    return bar(x) + bar(x + 1);
}

int main(void) {
    long x = 0;
    for (;;) {
        x = foo(x) % 1000;
    }
    return 0;
}

File: burn.py
#!/usr/bin/env python3
"""Small CPU-burning test program for exercising hamlet's profiler.

Runs a chain of nested function calls in a tight loop so the sampled
stack traces have a few recognizable frames (main -> foo -> bar).

Usage:
    python3 burn.py &
    cargo run --release -- --pid $!
"""


def bar(x):
    # Busy work with no syscalls, so the CPU stays pegged on this frame.
    total = 0
    for i in range(1000):
        total += (x * i) % 7919
    return total


def foo(x):
    return bar(x) + bar(x + 1)


def main():
    x = 0
    while True:
        x = foo(x) % 1000


if __name__ == "__main__":
    main()
